### ISSUE TITLE
Lei Mao's Log Book  Data Parallelism VS Model Parallelism in Distributed Deep Learning Training

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4890]
Lei Mao's Log Book – Data Parallelism VS Model Parallelism in Distributed Deep Learning Training<br>
<br>
https://ift.tt/2Yw42q3<br>
<br>
Twitter<br>
<br>
<div>
<div> <h3>Introduction</h3> <p>The number of parameters in modern deep learning models is becoming larger and larger, and the size of data set is also increasing dramatically. To train a sophisticated modern deep learning model on a large dataset, one has to use multi-node training otherwise it just takes forever. One may always see data parallelism and model parallelism in distributed deep learning training. In this blog post, I am going to talk about the theory, logic, and some misleading points about these two deep learning parallelism approaches.</p> <h3>Data Parallelism</h3> <p>In modern deep learning, because the dataset is too big to be fit into the memory, we could only do stochastic gradient descent for batches. For example, if we have 10K data points in the training dataset, every time we could only use 16 data points to calculate the estimate of the gradients, otherwise our GPU may stop working due to insufficient GPU memories.</p> <p><br></p> <p>The shortcoming of stochastic gradient descent is that the estimate of the gradients might not accurately represent the true gradients of using the full dataset. Therefore, it may take much longer to converge.</p> <p><br></p> <p>A natural way to have more accurate estimate of the gradients is to use larger batch sizes, or even use full dataset. To allow this, the gradients of small batches were calculated on each GPU, the final estimate of the gradients is the the weighted average of the gradients calculated from all the small batches.</p> <p><br></p> <p>Mathematically, data parallelism is valid because</p> <p>Where</p> <p><br></p> <p>$w$ is the parameters of the model,</p> <p>$\frac{\partial \text{ Loss}}{\partial w} $ is the true gradient of the big batch of size $n$,</p> <p>$\frac{\partial l_k}{\partial w}$ is the gradient of the small batch in GPU/node $k$,</p> <p>$x_i$ and $y_i$ are the features and labels of data point $i$,</p> <p>$f(x_i, y_i)$ is the loss for data point $i$ calculated from the forward propagation,</p> <p>$n$ is the total number of data points in the dataset,</p> <p>$k$ is the total number of GPUs/nodes,</p> <p>$m_k$ is the number of data points assigned to GPU/node $k$,</p> <p>$m_1 + m_2 + \cdots + m_k = n$.</p> <p><br></p> <p>When $m_1 = m_2 = \cdots = m_k = \frac{n}{k}$, we could further have</p> <p>Here for each GPU/node, we use the same model/parameters to do the forward propagation, we send a small batch of different data to each node, compute the gradient normally, and send the gradients back to the main node. This step is asynchronous because the speed of each GPU/node is slightly different. Once we got all the gradients (we are doing synchronization here), we calculate the (weighted) average of the gradients, and use the (weighted) average of the gradients to update the model/parameters. Then we move on to the next iteration.</p> <h3>Model Parallelism</h3> <p>Model parallelism sounds terrifying to me but it actually has nothing to do with math. It is a natural instinct of allocating computer resources. Sometimes we could not fit all the data into (GPU) memory is because there are too many layers and parameters in our deep learning model. Therefore we could divide the deep learning models to pieces, put a few consecutive layers to single node and calculate its gradients. In this way, the number of parameters in single node gets reduced and could use data to train to get more accurate gradients.</p> <p><br></p> <p>For example, we have 10 GPUs and we want to train a simple ResNet50 model. We could assign the first 5 layers to GPU #1, the second 5 layers to GPU #2, and so on, and the last 5 layers to GPU #10. During the training, in each iteration, the forward propagation has to be done in GPU #1 first. GPU #2 is waiting for the output from GPU #1, GPU #3 is waiting for the output from GPU #2, etc. Once the forward propagation is done. We calculate the gradients for the last layers which resides in GPU #10, and update the model parameters for those layers in GPU #10. Then the gradients back propagates to the previous layers in GPU #9, etc. Each GPU/node is like a compartment in the factory production line, it wait for the products from its previous compartment and send its own products to the next compartment.</p> <p>In my opinion, the name of model parallelism is misleading and it should not be considered as an example of parallel computing. A better name could be “Model Serialization”, since it is using a serial approach instead of a parallel approach in parallel computing. However, in some scenarios, some layers in some neural networks, such as <a href="https://leimao.github.io/article/Siamese-Network-MNIST/">Siamese Network</a>, are actually “parallel”. In that way, model parallelism could behave like real parallel computing to some extent. Data parallelism, however, is 100% parallel computing.</p> <p><br></p> <p>In addition, some interesting question, are some of the reinforcement learning algorithm such as A3C data parallelism or model parallelism. Well, I would say, A3C is more like data parallelism, although it is not exactly the same to the data parallelism described above since A3C does weight updates on the main node asynchronously throughout the whole training process while the process described above has clearly boundary of iterations and it does synchronization after iteration. You could disagree with me :)</p> <h3>References</h3> <hr>
<footer><div> <img alt="Lei Mao bio photo" src="https://leimao.github.io/images//author_images/Lei-Bio-Medium.jpg"><h3>Lei Mao</h3> <p>Machine Learning, Artificial Intelligence. On the Move.</p> <a href="http://twitter.com/matchaleimao" target="_blank"> Twitter</a> <a href="http://facebook.com/dukeleimao" target="_blank"><i></i> Facebook</a> <a href="http://linkedin.com/in/lei-mao" target="_blank"><i></i> LinkedIn</a> <a href="http://github.com/leimao" target="_blank"><i></i> GitHub</a> <a href="http://scholar.google.com/citations?user=R2VUf7YAAAAJ" target="_blank"><i></i>  G. Scholar</a> <a href="mailto:dukeleimao@gmail.com" target="_blank"><i></i> E-Mail</a> </div> <p><strong>Data Parallelism VS Model Parallelism in Distributed Deep Learning Training</strong> was published on <time datetime="2019-05-23T00:00:00-05:00">May 23, 2019</time> and last modified on <time datetime="2019-05-23">May 23, 2019</time> by <a title="About Lei Mao" href="https://leimao.github.io">Lei Mao</a>.</p> </footer>
</div> </div><br>
<br>
<br>
<br>
<br>
<br>
via Lei Mao's Log Book https://ift.tt/2MtcO1I<br>
<br>
July 31, 2019 at 10:56AM
